---
title: "COVID19 Sipros workflow"
author: "xyz"
date: "2021/5/31"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# install Sipros

```{bash}
conda create -n openmpi numpy scipy scikit-learn lxml python=2.7
conda activate openmpi
conda install -c conda-forge openmpi
git clone https://github.com/guo-xuan/Sipros-Ensemble.git
cd Sipros-Ensemble/
vim compiler.mk
# change following lines
# MCC := mpic++
# MGCC := mpicc
make openmp
```

# Preparing the protein databases 

[ensembl covid](ftp://ftp.ensemblgenomes.org/pub/viruses/fasta/sars_cov_2/pep/)
[ensembl Human CH38](http://ftp.ensembl.org/pub/release-104/fasta/homo_sapiens/pep/)
[cRAP protein sequences](https://www.thegpm.org/crap/)
[sars-cov-2 protein from uniprot](https://www.uniprot.org/proteomes/UP000464024)
[human protein from uniprot](https://www.uniprot.org/proteomes/UP000005640)
[Sindbis virus protein from uniprot](https://www.uniprot.org/proteomes/UP000006710)

```{bash}
# remove human proteins in cRAP pollution database
cd /mnt/d/ubuntuShare/COVID19/proteinDB
seqkit grep crap.fasta -p ".*HUMAN.*" -n -r -v > crap.NOhuman.fasta

# combine all proteins
cat crap.NOhuman.fasta \
  Sars-cov-2_uniprot-proteome_UP000464024.fasta \
  Sindbis_virus_uniprot-proteome_UP000006710.fasta\
  Human_uniprot-proteome_UP000005640.fasta \
  >ContaminationVirusHuman.fasta 
  
cd /scratch/yixiong/COVID19
cp /work/omicsbio/dywang/Guangpu/database_2.fasta protein_database.fasta
cd /scratch/yixiong/Sipros-Ensemble/Scripts/
python sipros_prepare_protein_database.py \
  -i /scratch/yixiong/COVID19/protein_database.fasta \
  -o /scratch/yixiong/COVID19/decoy.fasta \
  -c /scratch/yixiong/COVID19/SiprosConfig.cfg

python sipros_prepare_protein_database.py \
  -i /scratch/yixiong/COVID19/ContaminationVirusHuman.fasta \
  -o /scratch/yixiong/COVID19/ContaminationVirusHumanDecoy.fasta \
  -c /scratch/yixiong/COVID19/SiprosConfig.cfg
  
cd /scratch/yixiong/Sipros-Ensemble/bin
```

# convert raw to MS2

[proteowizard](http://proteowizard.sourceforge.net/download.html)
[proteowizard docker](https://hub.docker.com/r/chambm/pwiz-skyline-i-agree-to-the-vendor-licenses)

```{bash}
# docker version cannot work
cd /scratch/yixiong/COVID19/proteowizard
nohup cp /work/omicsbio/dywang/Guangpu/*.raw \
  ../RAW \
  > ../RAW/cp.log.txt 2>&1 & 
mkdir /mywineprefix
singularity pull --name skyline.simg \
  docker://chambm/pwiz-skyline-i-agree-to-the-vendor-licenses
singularity exec \
  --bind=../RAW/:/RAW,mywineprefix:/mywineprefix,../MS2:/MS2 \
  skyline.simg \
  mywine msconvert /RAW/Pan_050721_COVID2.raw \
  -o /MS2 --mzML --singleThreaded

# Thermo RAW reader not implemented: Thermo DLLs only work on Windows    
tar -jxvf pwiz-bin-linux-x86_64-gcc5-release-3_0_21149_88bef26.tar.bz2
./msconvert /work/omicsbio/dywang/Guangpu/*.raw \
  -o /scratch/yixiong/COVID19/ --ms2

cd "C:\green\proteowizard"
./msconvert /work/omicsbio/dywang/Guangpu/*.raw \
  -o /scratch/yixiong/COVID19/ --ms2
```

# run sipros

```{bash}
cd /scratch/yixiong/Sipros-Ensemble/bin
nohup ./Sipros_OpenMP -o /scratch/yixiong/COVID19/output \
  -w /scratch/yixiong/COVID19/MS2/MS2 \
  -c /scratch/yixiong/COVID19/SiprosConfig.cfg \
  > /scratch/yixiong/COVID19/Sipros.log.txt 2>&1 &
  
cd /scratch/yixiong/Sipros-Ensemble/Scripts 
nohup  ./runSiprosFiltering.sh \
  -in /scratch/yixiong/COVID19/output \
  -o /scratch/yixiong/COVID19/FilteringOutput \
  -c /scratch/yixiong/COVID19/SiprosConfig.cfg \
  > /scratch/yixiong/COVID19/Filtering.log.txt 2>&1 &
```

# run maxquant

```{bash}
conda install maxquant -c bioconda
cd /scratch/yixiong/COVID19/maxquant
# error: A required value not bound to option name is missing.
maxquant --changeFolder=/scratch/yixiong/COVID19/maxquant/mqparAllHumanProteins2.xml\
  /scratch/yixiong/COVID19/maxquant/proteinDB\
  /scratch/yixiong/COVID19/RAW\
  mqpar.xml
# convert file path in xml by hand
nohup maxquant mqparAllHumanProteins2.xml > maxquantLog.txt 2>&1 &
```

# deal with output of maxquant

Identification type: Indicates whether this experiment was identified by MS/MS 
or only by matching between runs.
  
Reverse: When marked with '+', this particular protein group contains noprotein,
made up of at least 50% of the peptides of the leading protein, 
with a peptide derived from the reversed part of the decoy database.
These should be removed for further data analysis. 

```{r}
library(data.table)
library(stringr)
library(VennDiagram)
library(hrbrthemes)
library(scales)
library(pheatmap)
df <- fread("proteinGroups.txt")
write.csv(df, "proteinGroups.csv", row.names = F)
# remove comtanmination
df <- df[!str_detect(df$`Majority protein IDs`, "CON__"), ]
df <- df[df$`Q-value` <= 0.01]
df <- df[df$Reverse != "+", ]
write.csv(df, "proteinGroupsFiltered.csv", row.names = F)
# identified 4084 protein groups
nrow(df)
# identified 17042 proteins
length(unlist(str_split(df$`Protein IDs`, ";")))

#### Venn diagram ####

COVID <- df$`Identification type COVID1` != "" |
  df$`Identification type COVID2` != "" |
  df$`Identification type COVID3` != ""
CTRL <- df$`Identification type ctrl1` != "" |
  df$`Identification type ctrl2` != "" |
  df$`Identification type ctrl3` != ""
S <- df$`Identification type S1` != "" |
  df$`Identification type S2` != "" |
  df$`Identification type S3` != ""
p <- venn.diagram(
  x = list(
    COVID = (1:nrow(df))[COVID],
    CTRL = (1:nrow(df))[CTRL],
    S = (1:nrow(df))[S]
  ),
  filename = NULL,
  output = TRUE ,
  # imagetype="png" ,
  height = 1200 ,
  width = 1200 ,
  resolution = 30,
  compression = "lzw",
  lwd = 1,
  col = c("#440154ff", '#21908dff', '#fde725ff'),
  fill = c(
    alpha("#440154ff", 0.3),
    alpha('#21908dff', 0.3),
    alpha('#fde725ff', 0.3)
  ),
  cex = 2,
  fontfamily = "sans",
  cat.cex = 2,
  cat.default.pos = "outer",
  cat.pos = c(-27, 27, 135),
  cat.dist = c(0.055, 0.055, 0.085),
  cat.fontfamily = "sans",
  cat.col = c("#440154ff", '#21908dff', '#fde725ff'),
  rotation = 1
)
pdf(file = "identified proteins venn.pdf")
grid.draw(p)
dev.off()

#### heatmap ####

mat <- as.matrix(df[, 65:73])
mat <- prop.table(mat,margin = 2)
# 4009 proteins have abundance
mat<-mat[rowSums(mat)!=0,]
mat <-scale(mat)
colnames(mat) <-
  c(paste0(rep(c(
    "COVID", "CTRL","S"
  ), each = 3), c(1:3, 1:3)))
colgroup <-
  data.frame(Treatments = factor(rep(c(
    "COVID", "CTRL","S"
  ), each = 3)))
row.names(colgroup) <- colnames(mat)

pdf("protein abundance heatmap with cluster.pdf",width=8,height = 6)
pheatmap(
  mat,
  show_rownames = F,
  fontsize = 15,
  # RdBu RdYlBu
  color = colorRampPalette(rev(
    RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")
  ))(100),
  cluster_row = T,
  cluster_col = T,
  scale = "row",
  annotation_col = colgroup
)
dev.off()

pdf("protein abundance heatmap without cluster.pdf",width=8,height = 6)
pheatmap(
  mat,
  show_rownames = F,
  fontsize = 15,
  # RdBu RdYlBu
  color = colorRampPalette(rev(
    RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")
  ))(100),
  cluster_row = T,
  cluster_col = F,
  scale = "row",
  annotation_col = colgroup
)
dev.off()
```

