---
title: "COVID19 Sipros workflow"
author: "xyz"
date: "2021/5/31"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# install Sipros

```{bash}
conda create -n openmpi numpy scipy scikit-learn lxml python=2.7
conda activate openmpi
conda install -c conda-forge openmpi
git clone https://github.com/guo-xuan/Sipros-Ensemble.git
cd Sipros-Ensemble/
vim compiler.mk
# change following lines
# MCC := mpic++
# MGCC := mpicc
make openmp
```

# Preparing the protein databases 

[ensembl covid](ftp://ftp.ensemblgenomes.org/pub/viruses/fasta/sars_cov_2/pep/)
[ensembl Human CH38](http://ftp.ensembl.org/pub/release-104/fasta/homo_sapiens/pep/)
[cRAP protein sequences](https://www.thegpm.org/crap/)
[sars-cov-2 protein from uniprot](https://www.uniprot.org/proteomes/UP000464024)
[human protein from uniprot](https://www.uniprot.org/proteomes/UP000005640)
[Sindbis virus protein from uniprot](https://www.uniprot.org/proteomes/UP000006710)

```{bash}
# remove human proteins in cRAP pollution database
cd /mnt/d/ubuntuShare/COVID19/proteinDB
seqkit grep crap.fasta -p ".*HUMAN.*" -n -r -v > crap.NOhuman.fasta

# combine all proteins
cat crap.NOhuman.fasta \
  Sars-cov-2_uniprot-proteome_UP000464024.fasta \
  Sindbis_virus_uniprot-proteome_UP000006710.fasta\
  Human_uniprot-proteome_UP000005640.fasta \
  >ContaminationVirusHuman.fasta 
  
cd /scratch/yixiong/COVID19
cp /work/omicsbio/dywang/Guangpu/database_2.fasta protein_database.fasta
cd /scratch/yixiong/Sipros-Ensemble/Scripts/
python sipros_prepare_protein_database.py \
  -i /scratch/yixiong/COVID19/protein_database.fasta \
  -o /scratch/yixiong/COVID19/decoy.fasta \
  -c /scratch/yixiong/COVID19/SiprosConfig.cfg

python sipros_prepare_protein_database.py \
  -i /scratch/yixiong/COVID19/ContaminationVirusHuman.fasta \
  -o /scratch/yixiong/COVID19/ContaminationVirusHumanDecoy.fasta \
  -c /scratch/yixiong/COVID19/SiprosConfig.cfg
  
cd /scratch/yixiong/Sipros-Ensemble/bin
```

# convert raw to MS2

[proteowizard](http://proteowizard.sourceforge.net/download.html)
[proteowizard docker](https://hub.docker.com/r/chambm/pwiz-skyline-i-agree-to-the-vendor-licenses)

```{bash}
# docker version cannot work
cd /scratch/yixiong/COVID19/proteowizard
nohup cp /work/omicsbio/dywang/Guangpu/*.raw \
  ../RAW \
  > ../RAW/cp.log.txt 2>&1 & 
mkdir /mywineprefix
singularity pull --name skyline.simg \
  docker://chambm/pwiz-skyline-i-agree-to-the-vendor-licenses
singularity exec \
  --bind=../RAW/:/RAW,mywineprefix:/mywineprefix,../MS2:/MS2 \
  skyline.simg \
  mywine msconvert /RAW/Pan_050721_COVID2.raw \
  -o /MS2 --mzML --singleThreaded

# Thermo RAW reader not implemented: Thermo DLLs only work on Windows    
tar -jxvf pwiz-bin-linux-x86_64-gcc5-release-3_0_21149_88bef26.tar.bz2
./msconvert /work/omicsbio/dywang/Guangpu/*.raw \
  -o /scratch/yixiong/COVID19/ --ms2

cd "C:\green\proteowizard"
./msconvert /work/omicsbio/dywang/Guangpu/*.raw \
  -o /scratch/yixiong/COVID19/ --ms2
```

# run sipros

```{bash}
cd /scratch/yixiong/Sipros-Ensemble/bin
nohup ./Sipros_OpenMP -o /scratch/yixiong/COVID19/output \
  -w /scratch/yixiong/COVID19/MS2/MS2 \
  -c /scratch/yixiong/COVID19/SiprosConfig.cfg \
  > /scratch/yixiong/COVID19/Sipros.log.txt 2>&1 &
  
cd /scratch/yixiong/Sipros-Ensemble/Scripts 
nohup  ./runSiprosFiltering.sh \
  -in /scratch/yixiong/COVID19/output \
  -o /scratch/yixiong/COVID19/FilteringOutput \
  -c /scratch/yixiong/COVID19/SiprosConfig.cfg \
  > /scratch/yixiong/COVID19/Filtering.log.txt 2>&1 &
```

# run maxquant

```{bash}
conda install maxquant -c bioconda
cd /scratch/yixiong/COVID19/maxquant
# error: A required value not bound to option name is missing.
maxquant --changeFolder=/scratch/yixiong/COVID19/maxquant/mqparAllHumanProteins2.xml\
  /scratch/yixiong/COVID19/maxquant/proteinDB\
  /scratch/yixiong/COVID19/RAW\
  mqpar.xml
# convert file path in xml by hand
nohup maxquant mqparAllHumanProteins2.xml > maxquantLog.txt 2>&1 &
```

# deal with output of maxquant

Identification type: Indicates whether this experiment was identified by MS/MS 
or only by matching between runs.
  
Reverse: When marked with '+', this particular protein group contains noprotein,
made up of at least 50% of the peptides of the leading protein, 
with a peptide derived from the reversed part of the decoy database.
These should be removed for further data analysis. 

```{r}
library(data.table)
library(stringr)
library(VennDiagram)
library(vegan)
library(ggplot2)
library(pheatmap)

# library(extrafont)
# fonts()
# # Show entire table
# fonttable()

df <- fread("proteinGroups.txt")
write.csv(df, "proteinGroups.csv", row.names = F)
# remove comtanmination
df <- df[!str_detect(df$`Majority protein IDs`, "CON__"), ]
df <- df[df$`Q-value` <= 0.01]
df <- df[df$Reverse != "+", ]
write.csv(df, "proteinGroupsFiltered.csv", row.names = F)
# identified 4084 protein groups
nrow(df)
# identified 17042 proteins
length(unlist(str_split(df$`Protein IDs`, ";")))

#### Venn diagram ####

COVID <- df$`Identification type COVID1` != "" |
  df$`Identification type COVID2` != "" |
  df$`Identification type COVID3` != ""
CTRL <- df$`Identification type ctrl1` != "" |
  df$`Identification type ctrl2` != "" |
  df$`Identification type ctrl3` != ""
S <- df$`Identification type S1` != "" |
  df$`Identification type S2` != "" |
  df$`Identification type S3` != ""
p <- venn.diagram(
  x = list(
    COVID = (1:nrow(df))[COVID],
    CTRL = (1:nrow(df))[CTRL],
    S = (1:nrow(df))[S]
  ),
  filename = NULL,
  output = TRUE ,
  # imagetype="png" ,
  height = 1200 ,
  width = 1200 ,
  resolution = 30,
  compression = "lzw",
  lwd = 1,
  col = c("#440154ff", '#21908dff', '#fde725ff'),
  fill = c(
    alpha("#440154ff", 0.3),
    alpha('#21908dff', 0.3),
    alpha('#fde725ff', 0.3)
  ),
  cex = 2,
  fontfamily = "sans",
  cat.cex = 2,
  cat.default.pos = "outer",
  cat.pos = c(-27, 27, 135),
  cat.dist = c(0.055, 0.055, 0.085),
  cat.fontfamily = "sans",
  cat.col = c("#440154ff", '#21908dff', '#fde725ff'),
  rotation = 1
)
pdf(file = "identified proteins venn.pdf")
grid.draw(p)
dev.off()

#### heatmap ####

mat <- df[, 65:73]
mat <-sapply(mat,as.numeric)
mat <- prop.table(mat,margin = 2)
# 4009 proteins have abundance
mat<-mat[rowSums(mat)!=0,]
colnames(mat) <-
  c(paste0(rep(c(
    "COVID", "CTRL","S"
  ), each = 3), c(1:3, 1:3)))
colgroup <-
  data.frame(Treatments = factor(rep(c(
    "COVID", "CTRL","S"
  ), each = 3)))
row.names(colgroup) <- colnames(mat)

pdf("protein abundance heatmap with cluster.pdf",width=8,height = 6)
pheatmap(
  mat,
  show_rownames = F,
  fontsize = 15,
  # RdBu RdYlBu
  color = colorRampPalette(rev(
    RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")
  ))(100),
  cluster_row = T,
  cluster_col = T,
  scale = "row",
  annotation_col = colgroup
)
dev.off()

pdf("protein abundance heatmap without cluster.pdf",width=8,height = 6)
pheatmap(
  mat,
  show_rownames = F,
  fontsize = 15,
  # RdBu RdYlBu
  color = colorRampPalette(rev(
    RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")
  ))(100),
  cluster_row = T,
  cluster_col = F,
  scale = "row",
  annotation_col = colgroup
)
dev.off()

# LFQ
# LFQ has more zero
sum(df[,84:92]==0);sum(df[,65:73]==0)

mat <- df[, 84:92]
mat <-sapply(mat,as.numeric)
mat <- prop.table(mat,margin = 2)
# 4009 proteins have abundance
mat<-mat[rowSums(mat)!=0,]
colnames(mat) <-
  c(paste0(rep(c(
    "COVID", "CTRL","S"
  ), each = 3), c(1:3, 1:3)))
colgroup <-
  data.frame(Treatments = factor(rep(c(
    "COVID", "CTRL","S"
  ), each = 3)))
row.names(colgroup) <- colnames(mat)

pdf("protein LFQ abundance heatmap with cluster.pdf",width=8,height = 6)
pheatmap(
  mat,
  show_rownames = F,
  fontsize = 15,
  # RdBu RdYlBu
  color = colorRampPalette(rev(
    RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")
  ))(100),
  cluster_row = T,
  cluster_col = T,
  scale = "row",
  annotation_col = colgroup
)
dev.off()

#### PCA ####

mat <- df[, 65:73]
mat <-sapply(mat,as.numeric)
mat <- prop.table(mat,margin = 2)
mat<-mat[rowSums(mat)!=0,]

pca <- rda(t(mat), scale = T)
importance <- summary(pca)[["cont"]][["importance"]]
siteScore <- summary(pca)[["sites"]]
tempDf <- data.frame(
  x = siteScore[, 1],
  y = siteScore[, 2],
  Treatments = colgroup$Treatments
)
ggplot(tempDf,
       aes(
         x = x,
         y = y,
         color = Treatments
       )) +
  geom_point(size = 3) +
  xlab(paste0("PC1(", round(importance[2, 1] * 100, 2), "%)")) +
  ylab(paste0("PC2(", round(importance[2, 2] * 100, 2), "%)")) +
  theme(text = element_text(size = 20)) +
  ggsave(
    "protein abundance PCA.pdf",
    width = 8,
    height = 6
  )
```

# protein with different abundance

```{r}
library(DESeq2)
library(ggbreak)

#### COVID vs CTRL ####
mat <- df[, 65:70]
mat <-sapply(mat,as.numeric)
# some is too large and overflow integer
mat <- round(mat/1000)
row.names(mat)<-df$`Majority protein IDs`
mat<-mat[rowSums(mat)!=0,]
coldata <- data.frame(row.names = colnames(mat),
                      condition = factor(colgroup$Treatments[1:6]))
dds <- DESeqDataSetFromMatrix(countData = mat,
                              colData = coldata,
                              design =  ~ condition)
dds2 <- DESeq(dds)
DEGs <- results(dds2, contrast=c("condition","COVID","CTRL"))
DEGsDf <- na.exclude(as.data.frame(DEGs))
saveRDS(DEGsDf,"COVID vs CTRL.rds")
write.csv(DEGsDf,"COVID vs CTRL.csv")
DEGsDf <-
  readRDS("COVID vs CTRL.rds")
# The Standard of significance was fold change >= 2,
# or =< 1/2 (log(DEGsDlog2FoldChangef)>=1 or <=-1),
# adjusted P value =< 0.05
judgeSignificance <- function(x, y) {
  if (x <= 0.05) {
    if (y <= -1)
      return ("down")
    else if (y >= 1)
      return ("up")
    else
      return ("no")
  }
  else
    return("no")
}
DEGsDf$significance<-mapply(judgeSignificance, DEGsDf$padj, DEGsDf$log2FoldChange)
DEGsDfFiltered<-DEGsDf[DEGsDf$significance!="no",]
write.csv(DEGsDfFiltered,"Protein with different abundance COVID vs CTRL.csv")
# minimal padj is 0
head(sort(log10(DEGsDf$padj)))
# change to none zero minimal
DEGsDf$padj[DEGsDf$padj==0]<-min(DEGsDf$padj[DEGsDf$padj>0])
drawVolcanoPlot <- function(df) {
  colnames(df) <- c("FDR", "FC")
  df$size <- -log10(df$FDR)
  df <-
    cbind(df, significance = factor(mapply(judgeSignificance, df$FDR, df$FC)))
  ggplot(df, aes(FC, -1 * log10(FDR))) +
    geom_point(
      aes(
        fill = significance,
        colour = significance,
        size = size
      ),
      shape = 21,
      alpha = 0.6
    ) +
    # Resize the point from 0.1 to 10
    scale_size(range = c(.1, 10)) +
    theme(text = element_text(size = 20)) +
    guides(
      fill = guide_legend(override.aes = list(size = 10)),
      size = FALSE,
      colour = FALSE
    ) +
    scale_fill_discrete(labels = paste(levels(df$significance), table(df$significance))) +
    labs(x = expression(log[2](FC)), y = expression(-log[10](FDR))) +
    geom_hline(yintercept = -log10(0.05), linetype = 4) +
    geom_vline(xintercept = c(-log2(2), log2(2)), linetype = 4) +
    # Limit the scope of the drawing, make it left-right symmetric, and add linear coordinates
    scale_x_continuous(limits = c(-31, 31), breaks = sort(c(seq(
      -30,
      30, length.out = 6
    ),-1, 1)))
}

p<-drawVolcanoPlot(DEGsDf[, c("padj", "log2FoldChange")], )
# Limit the scope of the drawing and add linear coordinates
p<-p+ scale_y_continuous(limits = c(0, 95), breaks = c(1.3, 10, 20, 40, 80))
p+ggsave("COVID vs CTRL volcano plot.pdf",width = 8,height = 6)

#### S vs CTRL ####

mat <- df[, c(71:73,68:70)]
mat <-sapply(mat,as.numeric)
# some is too large and overflow integer
mat <- round(mat/1000)
row.names(mat)<-df$`Majority protein IDs`
mat<-mat[rowSums(mat)!=0,]
coldata <- data.frame(row.names = colnames(mat),
                      condition = factor(colgroup$Treatments[c(7:9,4:6)]))
dds <- DESeqDataSetFromMatrix(countData = mat,
                              colData = coldata,
                              design =  ~ condition)
dds2 <- DESeq(dds)
DEGs <- results(dds2, contrast=c("condition","S","CTRL"))
DEGsDf <- na.exclude(as.data.frame(DEGs))
saveRDS(DEGsDf,"S vs CTRL.rds")
write.csv(DEGsDf,"S vs CTRL.csv")
DEGsDf <-
  readRDS("S vs CTRL.rds")
DEGsDf$significance<-mapply(judgeSignificance, DEGsDf$padj, DEGsDf$log2FoldChange)
DEGsDfFiltered<-DEGsDf[DEGsDf$significance!="no",]
write.csv(DEGsDfFiltered,"Protein with different abundance S vs CTRL.csv")
head(sort(log10(DEGsDf$padj)))
DEGsDf$padj[DEGsDf$padj==0]<-min(DEGsDf$padj[DEGsDf$padj>0])
p<-drawVolcanoPlot(DEGsDf[, c("padj", "log2FoldChange")])
# Limit the scope of the drawing and add linear coordinates
p<-p+ scale_y_continuous(limits = c(0, 162), breaks = c(1.3, 10, 20, 40, 80, 160))
# scale_y_break conflict with geom_hline and geom_vline
# p<-p+ scale_y_break(c(40, 150))
p+ggsave("S vs CTRL volcano plot.pdf",width = 8,height = 6)
```

# get protein annotation

```{r}
library(AnnotationHub)
library(clusterProfiler)
library(stringr)

# refreshHub(hubClass="AnnotationHub")
ah <- AnnotationHub()
# job::job({display(ah)})
# Homo sapiens OrgDb
hm <- ah[["AH92581"]]
hmDf <-AnnotationDbi::select(hm, keys=keys(hm), columns = c("UNIPROT","GO","PATH","ENZYME"))
hmDf <- hmDf[!is.na(hmDf$UNIPROT),]
goName <-
  AnnotationDbi::select(
    x = GO.db::GO.db,
    keys = hmDf$GO,
    keytype = "GOID",
    columns = "TERM"
  )

#### COVID ####

covid <-read.csv("Protein with different abundance COVID vs CTRL.csv")
covidList<-unlist(str_split(covid$X,";"))
covidList<-unlist(str_split(covidList,"\\|",simplify = T)[,2])

goMF <- enrichGO(gene = covidList,OrgDb = hm,keyType = "UNIPROT",ont = "MF",pvalueCutoff=0.05)
goMFdf <- as.data.frame(goMF)
write.csv(goMFdf,"COVID vs CTRL MF enrich.csv",row.names = F)
p<-dotplot(goMF,font.size=10)
p+ggsave("COVID vs CTRL MF enrich.pdf",width = 8,height = 6)

goBP <- enrichGO(gene = covidList,OrgDb = hm,keyType = "UNIPROT",ont = "BP",pvalueCutoff=0.05)
goBPdf <- as.data.frame(goBP)
write.csv(goBPdf,"COVID vs CTRL BP enrich.csv",row.names = F)
p<-dotplot(goBP,font.size=10)
p+ggsave("COVID vs CTRL BP enrich.pdf",width = 8,height = 6)

goCC <- enrichGO(gene = covidList,OrgDb = hm,keyType = "UNIPROT",ont = "CC",pvalueCutoff=0.05)
goCCdf <- as.data.frame(goCC)
write.csv(goCCdf,"COVID vs CTRL CC enrich.csv",row.names = F)
p<-dotplot(goCC,font.size=10)
p+ggsave("COVID vs CTRL CC enrich.pdf",width = 8,height = 6)

kegg <- enrichKEGG(gene = covidList,organism = "hsa",keyType = "uniprot",pvalueCutoff=0.05)
keggDf <- as.data.frame(kegg)
write.csv(keggDf,"COVID vs CTRL KEGG enrich.csv",row.names = F)
p<-dotplot(kegg,font.size=10)
p+ggsave("COVID vs CTRL KEGG enrich.pdf",width = 8,height = 6)

#### S ####

S <-read.csv("Protein with different abundance S vs CTRL.csv")
SList<-unlist(str_split(S$X,";"))
SList<-unlist(str_split(SList,"\\|",simplify = T)[,2])

goMF <- enrichGO(gene = SList,OrgDb = hm,keyType = "UNIPROT",ont = "MF",pvalueCutoff=0.05)
goMFdf <- as.data.frame(goMF)
write.csv(goMFdf,"S vs CTRL MF enrich.csv",row.names = F)
p<-dotplot(goMF,font.size=10)
p+ggsave("S vs CTRL MF enrich.pdf",width = 8,height = 6)

goBP <- enrichGO(gene = SList,OrgDb = hm,keyType = "UNIPROT",ont = "BP",pvalueCutoff=0.05)
goBPdf <- as.data.frame(goBP)
write.csv(goBPdf,"S vs CTRL BP enrich.csv",row.names = F)
p<-dotplot(goBP,font.size=10)
p+ggsave("S vs CTRL BP enrich.pdf",width = 8,height = 6)

goCC <- enrichGO(gene = SList,OrgDb = hm,keyType = "UNIPROT",ont = "CC",pvalueCutoff=0.05)
goCCdf <- as.data.frame(goCC)
write.csv(goCCdf,"S vs CTRL CC enrich.csv",row.names = F)
p<-dotplot(goCC,font.size=10)
p+ggsave("S vs CTRL CC enrich.pdf",width = 8,height = 6)

kegg <- enrichKEGG(gene = SList,organism = "hsa",keyType = "uniprot",pvalueCutoff=1)
keggDf <- as.data.frame(kegg)
write.csv(keggDf,"S vs CTRL KEGG enrich.csv",row.names = F)
```

# gseGO

```{r}
library(tidyr)
covid <-readRDS("COVID vs CTRL.rds")
covid$padj[covid$padj==0]<-min(covid$padj[covid$padj>0])
covid <- data.frame(name=row.names(covid),p=-log10(covid$padj))
covid <- separate_rows(covid,name,sep = ";")
covid$name <- str_split(covid$name,"\\|",simplify = T)[,2]
covid <- arrange(covid,desc(p))
proteinList<-covid$p
names(proteinList)<-covid$name
goMF <- gseGO(gene = proteinList,OrgDb = hm,keyType = "UNIPROT",ont = "MF")
dotplot(goMF,font.size=10)
```

# sipros result

```{r}
library(data.table)
library(ggvenn)
siDf<-fread("Pan_050721.pro.txt",skip = 58)
si<-siDf$ProteinID
si<-si[!str_detect(si,"Rev2_")]
si<-str_replace(si,"\\{","")
si<-str_replace(si,"\\}","")
si<-unlist(str_split(si,","))
mq<-read.csv("proteinGroupsFiltered.csv")
mq<-mq$Majority.protein.IDs
mq<-unlist(str_split(mq,";"))
sum(si %in% mq)
p <- venn.diagram(
  x = list(
    Sipros = si,
    Maxquant = mq
  ),
  filename = NULL,
  output = F ,
  height = 1200 ,
  width = 1200 ,
  resolution = 30
)
pdf(file = "Sipros Maxquant venn.pdf")
grid.draw(p)
dev.off()

a <- list(
    Sipros = si,
    Maxquant = mq
  )
p<-ggvenn(a, c("Sipros", "Maxquant"))
p+ggsave("Sipros Maxquant venn2.pdf")
```

