---
title: "use green monkey"
author: "xyz"
date: "2021/7/14"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# database

Quantitative Proteomics Reveals Changes in Vero Cells in Response to Porcine Epidemic Diarrhea Virus

Protein identification and quantitation were performed with MaxQuant (Max-Planck-Institute of Biochemistry, version 1.1.1.2) and Mascot (Matrix Science, version 2.2) by searching against a concatenated forward–reverse database from the UniProt database (http://www.uniprot.org/taxonomy/9528, November 12, 2013, containing 148 822 sequences) and PEDV prototype strain CV777 (http://www.uniprot.org/taxonomy/229032, January 8, 2014, containing seven entities).


Quantitative Proteomics Using Stable Isotope Labeling with Amino Acids in Cell Culture Reveals Changes in the Cytoplasmic, Nuclear, and Nucleolar Proteomes in Vero Cells Infected with the Coronavirus Infectious Bronchitis Virus*

The derived peak list was searched with the Mascot search engine (version 2.1.04; Matrix Science, London, UK) against a concatenated database combining 80,412 proteins from the International Protein Index (IPI) human protein database version 3.6 (forward database) and the reversed sequences of all proteins (reverse database). 


Proteome analysis of porcine epidemic diarrhea virus (PEDV)-infected Vero cells

The acquired MS raw data files were converted into MGF files by using 5600 ms converter and the MGF files were further searched. MS/MS data were searched against the monkey sequence database from the NCBInr database (release April 2013, containing 29 736 sequences) and the proteins were identified using MASCOT search engine (Matrix Science, London, UK; version 2.3.02).

[Chlorocebus sabaeus (Green monkey)](https://www.uniprot.org/proteomes/UP000029965)
[Chlorocebus sabaeus Vervet-AGM Ensembl](https://asia.ensembl.org/Chlorocebus_sabaeus/Info/Index)

format  type     num_seqs     sum_len  min_len  avg_len  max_len
FASTA   Protein    19,229  10,437,685       18    542.8   35,229

```{bash}
seqkit stat Green_Monkey_uniprot-proteome_UP000029965.fasta 
seqkit seq Green_Monkey_uniprot-proteome_UP000029965.fasta \
  '../only use Swiss-prot proteins/uniprot-proteome_UP000006710+reviewed_yes.fasta' \
  '../only use Swiss-prot proteins/uniprot-proteome_UP000464024+reviewed_yes.fasta' \
  > monkeyCOVIDdb.fasta
  
cd /scratch/yixiong/COVID19/maxquantOutPut20210716
conda activate openmpi
nohup maxquant monkeyCOVID.xml > maxquantLog.txt 2>&1 &
```

# deal with output of maxquant

Identification type: Indicates whether this experiment was identified by MS/MS 
or only by matching between runs.
  
Reverse: When marked with '+', this particular protein group contains noprotein,
made up of at least 50% of the peptides of the leading protein, 
with a peptide derived from the reversed part of the decoy database.
These should be removed for further data analysis. 

```{r}
library(data.table)
library(stringr)
library(VennDiagram)
library(vegan)
library(ggplot2)
library(scales)
library(pheatmap)
library(ggvenn)
library(AnnotationHub)
library(clusterProfiler)
library(stringr)
library(plyr)
library(dplyr)
library(tidyr)
library(DESeq2)
library(AnnotationForge)
library(RSQLite)
library(biomaRt)
library(KEGGREST)
```

```{r}
df <- fread("proteinGroups.txt")
write.csv(df, "proteinGroups.csv", row.names = F)
# remove comtanmination
df <- df[!str_detect(df$`Majority protein IDs`, "CON__"), ]
df <- df[df$`Q-value` <= 0.01]
df <- df[df$Reverse != "+", ]
write.csv(df, "proteinGroupsFiltered.csv", row.names = F)
# identified 4252 protein groups
df <- read.csv("proteinGroupsFiltered.csv", check.names = F)
nrow(df)
# identified 4710 proteins
length(unlist(str_split(df$`Protein IDs`, ";")))

#### Venn diagram ####

COVID <- df$`Identification type COVID1` == "By MS/MS" |
  df$`Identification type COVID2` == "By MS/MS" |
  df$`Identification type COVID3` == "By MS/MS"
CTRL <- df$`Identification type ctrl1` == "By MS/MS" |
  df$`Identification type ctrl2` == "By MS/MS" |
  df$`Identification type ctrl3` == "By MS/MS"
S <- df$`Identification type S1` == "By MS/MS" |
  df$`Identification type S2` == "By MS/MS" |
  df$`Identification type S3` == "By MS/MS"
pN <- 1:nrow(df)
pList <- list(COVID = pN[COVID],
              CTRL = pN[CTRL],
              S = pN[S])
p <- ggvenn(pList)
p + ggsave("identified proteins venn.pdf",
           width = 8,
           height = 6)

#### if MS2 count==0 convert intensity and LFQ intensity to 0 ####

isZero <- df[, 93:101] == 0
df[, 65:73][isZero] <- 0
df[, 84:92][isZero] <- 0

write.csv(df, "proteinGroupsFilteredFixed.csv", row.names = F)

#### heatmap ####

mat <- df[, 65:73]
mat <- sapply(mat, as.numeric)
mat <- prop.table(mat, margin = 2)
# 4181 proteins have abundance
mat <- mat[rowSums(mat) != 0,]
colnames(mat) <-
  c(paste0(rep(c("COVID", "CTRL", "S"), each = 3), c(1:3, 1:3)))
colgroup <-
  data.frame(Treatments = factor(rep(c("COVID", "CTRL", "S"), each = 3)))
row.names(colgroup) <- colnames(mat)

pdf("protein intensity heatmap with cluster.pdf",
    width = 8,
    height = 6)
pheatmap(
  mat,
  show_rownames = F,
  fontsize = 15,
  # RdBu RdYlBu
  color = colorRampPalette(rev(
    RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")
  ))(100),
  cluster_row = T,
  cluster_col = T,
  scale = "row",
  annotation_col = colgroup
)
dev.off()

pdf("protein abundance heatmap without cluster.pdf",
    width = 8,
    height = 6)
pheatmap(
  mat,
  show_rownames = F,
  fontsize = 15,
  # RdBu RdYlBu
  color = colorRampPalette(rev(
    RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")
  ))(100),
  cluster_row = T,
  cluster_col = F,
  scale = "row",
  annotation_col = colgroup
)
dev.off()

# LFQ
# LFQ has more zero
sum(df[, 84:92] == 0)
sum(df[, 65:73] == 0)

mat <- df[, 84:92]
mat <- sapply(mat, as.numeric)
mat <- prop.table(mat, margin = 2)
# 4009 proteins have abundance
mat <- mat[rowSums(mat) != 0,]
colnames(mat) <-
  c(paste0(rep(c("COVID", "CTRL", "S"), each = 3), c(1:3, 1:3)))
colgroup <-
  data.frame(Treatments = factor(rep(c("COVID", "CTRL", "S"), each = 3)))
row.names(colgroup) <- colnames(mat)

pdf("protein LFQ abundance heatmap with cluster.pdf",
    width = 8,
    height = 6)
pheatmap(
  mat,
  show_rownames = F,
  fontsize = 15,
  # RdBu RdYlBu
  color = colorRampPalette(rev(
    RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")
  ))(100),
  cluster_row = T,
  cluster_col = T,
  scale = "row",
  annotation_col = colgroup
)
dev.off()

# MS2 count

mat <- df[, 93:101]
mat <- sapply(mat, as.numeric)
mat <- mat[rowSums(mat) != 0,]
mat <- prop.table(mat, 2)
colnames(mat) <-
  c(paste0(rep(c("COVID", "CTRL", "S"), each = 3), c(1:3, 1:3)))
colgroup <-
  data.frame(Treatments = factor(rep(c("COVID", "CTRL", "S"), each = 3)))
row.names(colgroup) <- colnames(mat)

pdf("protein MS2 count heatmap.pdf",
    width = 8,
    height = 6)
pheatmap(
  mat,
  show_rownames = F,
  fontsize = 15,
  # RdBu RdYlBu
  color = colorRampPalette(rev(
    RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")
  ))(100),
  cluster_row = T,
  cluster_col = T,
  scale = "row",
  annotation_col = colgroup
)

#### PCA ####

mat <- df[, 65:73]
mat <- sapply(mat, as.numeric)
mat <- prop.table(mat, margin = 2)
mat <- mat[rowSums(mat) != 0, ]

pca <- rda(t(mat), scale = T)
importance <- summary(pca)[["cont"]][["importance"]]
siteScore <- summary(pca)[["sites"]]
tempDf <- data.frame(x = siteScore[, 1],
                     y = siteScore[, 2],
                     Treatments = colgroup$Treatments)
ggplot(tempDf,
       aes(x = x,
           y = y,
           color = Treatments)) +
  geom_point(size = 3) +
  xlab(paste0("PC1(", round(importance[2, 1] * 100, 2), "%)")) +
  ylab(paste0("PC2(", round(importance[2, 2] * 100, 2), "%)")) +
  theme(text = element_text(size = 20)) +
  ggsave("protein abundance PCA.pdf",
         width = 8,
         height = 6)

mat <- df[, 84:92]
mat <- sapply(mat, as.numeric)
mat <- prop.table(mat, margin = 2)
mat <- mat[rowSums(mat) != 0, ]

pca <- rda(t(mat), scale = T)
importance <- summary(pca)[["cont"]][["importance"]]
siteScore <- summary(pca)[["sites"]]
tempDf <- data.frame(x = siteScore[, 1],
                     y = siteScore[, 2],
                     Treatments = colgroup$Treatments)
ggplot(tempDf,
       aes(x = x,
           y = y,
           color = Treatments)) +
  geom_point(size = 3) +
  xlab(paste0("PC1(", round(importance[2, 1] * 100, 2), "%)")) +
  ylab(paste0("PC2(", round(importance[2, 2] * 100, 2), "%)")) +
  theme(text = element_text(size = 20)) +
  ggsave("protein LFQ abundance PCA.pdf",
         width = 8,
         height = 6)
```

# protein with different abundance

```{r}
#### COVID vs CTRL ####
mat <- df[, 65:70]
mat <- sapply(mat, as.numeric)
# some is too large and overflow integer
mat <- round(mat / 1000)
row.names(mat) <- df$`Majority protein IDs`
mat <- mat[rowSums(mat) != 0, ]
coldata <- data.frame(row.names = colnames(mat),
                      condition = factor(colgroup$Treatments[1:6]))
dds <- DESeqDataSetFromMatrix(countData = mat,
                              colData = coldata,
                              design =  ~ condition)
dds2 <- DESeq(dds, fitType = 'local')
DEGs <- results(dds2, contrast = c("condition", "COVID", "CTRL"))
DEGsDf <- na.exclude(as.data.frame(DEGs))
saveRDS(DEGsDf, "COVID vs CTRL.rds")
write.csv(DEGsDf, "COVID vs CTRL.csv")
DEGsDf <-
  readRDS("COVID vs CTRL.rds")
# The Standard of significance was fold change >= 2,
# or =< 1/2 (log(DEGsDlog2FoldChangef)>=1 or <=-1),
# adjusted P value =< 0.05
judgeSignificance <- function(x, y) {
  if (x <= 0.05) {
    if (y <= -1)
      return ("down")
    else if (y >= 1)
      return ("up")
    else
      return ("no")
  }
  else
    return("no")
}
DEGsDf$significance <-
  mapply(judgeSignificance, DEGsDf$padj, DEGsDf$log2FoldChange)
DEGsDfFiltered <- DEGsDf[DEGsDf$significance != "no", ]
write.csv(DEGsDfFiltered,
          "Protein with different abundance COVID vs CTRL.csv")
# minimal padj is 0
head(sort(log10(DEGsDf$padj)))
# change to none zero minimal
DEGsDf$padj[DEGsDf$padj == 0] <- min(DEGsDf$padj[DEGsDf$padj > 0])
drawVolcanoPlot <- function(df) {
  colnames(df) <- c("FDR", "FC")
  df$size <- -log10(df$FDR)
  df <-
    cbind(df, significance = factor(mapply(judgeSignificance, df$FDR, df$FC)))
  ggplot(df, aes(FC,-1 * log10(FDR))) +
    geom_point(
      aes(
        fill = significance,
        colour = significance,
        size = size
      ),
      shape = 21,
      alpha = 0.6
    ) +
    # Resize the point from 0.1 to 10
    scale_size(range = c(.1, 10)) +
    theme(text = element_text(size = 20)) +
    guides(
      fill = guide_legend(override.aes = list(size = 10)),
      size = FALSE,
      colour = FALSE
    ) +
    scale_fill_discrete(labels = paste(levels(df$significance), table(df$significance))) +
    labs(x = expression(log[2](FC)), y = expression(-log[10](FDR))) +
    geom_hline(yintercept = -log10(0.05), linetype = 4) +
    geom_vline(xintercept = c(-log2(2), log2(2)), linetype = 4) +
    # Limit the scope of the drawing, make it left-right symmetric, and add linear coordinates
    scale_x_continuous(limits = c(-31, 31), breaks = sort(c(seq(
      -30,
      30, length.out = 6
    ), -1, 1)))
}

p <- drawVolcanoPlot(DEGsDf[, c("padj", "log2FoldChange")])
# Limit the scope of the drawing and add linear coordinates
p <-
  p + scale_y_continuous(limits = c(0, 200), breaks = c(1.3, 10, 20, 40, 80, 160))
p + ggsave("COVID vs CTRL volcano plot.pdf",
           width = 8,
           height = 6)

#### S vs CTRL ####

mat <- df[, c(71:73, 68:70)]
mat <- sapply(mat, as.numeric)
# some is too large and overflow integer
mat <- round(mat / 1000)
row.names(mat) <- df$`Majority protein IDs`
mat <- mat[rowSums(mat) != 0, ]
coldata <- data.frame(row.names = colnames(mat),
                      condition = factor(colgroup$Treatments[c(7:9, 4:6)]))
dds <- DESeqDataSetFromMatrix(countData = mat,
                              colData = coldata,
                              design =  ~ condition)
dds2 <- DESeq(dds,fitType='local')
DEGs <- results(dds2, contrast = c("condition", "S", "CTRL"))
DEGsDf <- na.exclude(as.data.frame(DEGs))
saveRDS(DEGsDf, "S vs CTRL.rds")
write.csv(DEGsDf, "S vs CTRL.csv")
DEGsDf <-
  readRDS("S vs CTRL.rds")
DEGsDf$significance <-
  mapply(judgeSignificance, DEGsDf$padj, DEGsDf$log2FoldChange)
DEGsDfFiltered <- DEGsDf[DEGsDf$significance != "no", ]
write.csv(DEGsDfFiltered, "Protein with different abundance S vs CTRL.csv")
head(sort(log10(DEGsDf$padj)))
DEGsDf$padj[DEGsDf$padj == 0] <- min(DEGsDf$padj[DEGsDf$padj > 0])
p <- drawVolcanoPlot(DEGsDf[, c("padj", "log2FoldChange")])
# Limit the scope of the drawing and add linear coordinates
p <-
  p + scale_y_continuous(limits = c(0, 123),
                         breaks = c(1.3, 10, 20, 40, 80, 130))
# scale_y_break conflict with geom_hline and geom_vline
# p<-p+ scale_y_break(c(40, 150))
p + ggsave("S vs CTRL volcano plot.pdf",
           width = 8,
           height = 6)

#### COVID vs S ####

mat <- df[,c(65:67, 71:73)]
mat <- sapply(mat, as.numeric)
# some is too large and overflow integer
mat <- round(mat / 1000)
row.names(mat) <- df$`Majority protein IDs`
mat <- mat[rowSums(mat) != 0, ]
coldata <- data.frame(row.names = colnames(mat),
                      condition = factor(colgroup$Treatments[c(1:3,7:9)]))
dds <- DESeqDataSetFromMatrix(countData = mat,
                              colData = coldata,
                              design =  ~ condition)
dds2 <- DESeq(dds,fitType='local')
DEGs <- results(dds2, contrast = c("condition", "COVID", "S"))
DEGsDf <- na.exclude(as.data.frame(DEGs))
saveRDS(DEGsDf, "COVID vs S.rds")
write.csv(DEGsDf, "COVID vs S.csv")
DEGsDf <-
  readRDS("COVID vs S.rds")
judgeSignificance <- function(x, y) {
  if (x <= 0.05) {
    if (y <= -1)
      return ("down")
    else if (y >= 1)
      return ("up")
    else
      return ("no")
  }
  else
    return("no")
}
DEGsDf$significance <-
  mapply(judgeSignificance, DEGsDf$padj, DEGsDf$log2FoldChange)
DEGsDfFiltered <- DEGsDf[DEGsDf$significance != "no", ]
write.csv(DEGsDfFiltered,
          "Protein with different abundance COVID vs S.csv")
# minimal padj is 0
head(sort(log10(DEGsDf$padj)))
# change to none zero minimal
DEGsDf$padj[DEGsDf$padj == 0] <- min(DEGsDf$padj[DEGsDf$padj > 0])
p <- drawVolcanoPlot(DEGsDf[, c("padj", "log2FoldChange")])
p <-
  p + scale_y_continuous(limits = c(0, 130), breaks = c(1.3, 10, 20, 40, 80, 130))
p + ggsave("COVID vs S volcano plot.pdf",
           width = 8,
           height = 6)
```

# use biomart download annotation

[biomart webpage](http://uswest.ensembl.org/biomart/martview/)

```{r}
listMarts()
ensembl.base <- useMart(biomart = "ENSEMBL_MART_ENSEMBL")
df <- listDatasets(ensembl.base)
GMbase <-
  useDataset(mart = ensembl.base, dataset = "csabaeus_gene_ensembl")
View(listAttributes(GMbase))
# UniProtKB/TrEMBL ID, Gene name, GO term accession, GO term name, GO term definition, GO term evidence code, GO domain
geneName <-
  getBM(
    attributes = c("uniprotsptrembl",
                   "uniprot_gn_symbol",
                   "description"),
    mart = GMbase
  )
geneName <- geneName[geneName$uniprotsptrembl != "", ]
saveRDS(geneName, "geneName.rds")
GO <-
  getBM(
    attributes = c(
      "uniprotsptrembl",
      "go_id",
      "name_1006",
      "definition_1006",
      "go_linkage_type",
      "namespace_1003"
    ),
    mart = GMbase
  )
GO <- GO[GO$uniprotsptrembl != "", ]
colnames(GO)[-1] <-
  c("GO_ID",
    "GO_Term",
    "GO_Term_Definition",
    "GO_Evidence_Code",
    "GO_Domain")
saveRDS(GO, "GO.rds")
kegg <-
  getBM(attributes = c("kegg_enzyme",
                       "uniprotsptrembl"),
        mart = GMbase)
kegg <- kegg[kegg$uniprotsptrembl != "", ]
kegg <- kegg[!is.na(kegg$kegg_enzyme), ]
kegg$kegg_enzyme <- sprintf("K%05d", kegg$kegg_enzyme)
K04714<-kegg[kegg$kegg_enzyme=="K04714",]
# 7086 protein has K04714, kegg annotation is not accurate
length(unique(K04714$uniprotsptrembl))
colnames(kegg)[2] <- "KEGG_KO"
KO <- keggList("ko")
KOdf <- data.frame(id = names(KO), term = KO)
KOdf$id <- str_sub(KOdf$id, 4)
KOtoPathway <- keggLink("ko", "pathway")
KOtoPathway <- data.frame(ko = str_sub(KOtoPathway, 4),
                          path = str_sub(names(KOtoPathway), 6))
KOtoPathway <- KOtoPathway[!str_detect(KOtoPathway$path, "ko"), ]
pathway <- keggList("pathway")
pathwaydf <- data.frame(id = names(pathway), term = pathway)
pathwaydf$id <- str_sub(pathwaydf$id, 6)
keggAnnotation <- left_join(KOdf, KOtoPathway, by = c("id" = "ko"))
keggAnnotation <-
  left_join(keggAnnotation, pathwaydf, by = c("path" = "id"))
colnames(keggAnnotation) <- c("KO", "KO_TERM", "PATH", "PATH_TERM")
saveRDS(keggAnnotation, "keggAnnotation.rds")
gmKegg <- left_join(kegg, keggAnnotation, by = c("KEGG_KO" = "KO"))
saveRDS(gmKegg, "gmKegg.rds")
```

## add annotation to protein

```{r}
geneName<-readRDS("geneName.rds")
geneName <- as_tibble(geneName)
geneName <- group_by(geneName, uniprotsptrembl)
# merge rows by same value in one column
pasteSemi <- function(x)
  paste(unique(x), collapse = " | ")
geneName <-
  summarise(geneName, across(uniprot_gn_symbol:description, pasteSemi))

GO<-readRDS("GO.rds")
GO <- as_tibble(GO)
GO <- group_by(GO, uniprotsptrembl)
pasteSemi <- function(x)
  paste(unique(x), collapse = " | ")
GO <-
  summarise(GO, across(GO_ID:GO_Domain, pasteSemi))

kegg<-readRDS("gmKegg.rds")
kegg <- as_tibble(kegg)
kegg <- group_by(kegg, uniprotsptrembl)
pasteSemi <- function(x)
  paste(unique(x), collapse = " | ")
kegg <-
  dplyr::summarise(kegg, across(KEGG_KO:PATH_TERM, pasteSemi))

# annoDf<-join_all(list(geneName=geneName,GO=GO,kegg=kegg),by="uniprotsptrembl")
annoDf<-join_all(list(geneName=geneName,GO=GO),by="uniprotsptrembl")

addAnnotation <- function(filePath) {
  pDF <-
    read.csv(filePath, check.names = F)
  colnames(pDF)[1] <- "ID"
  pDF <- separate_rows(pDF, ID, sep = ";")
  id <- unlist(str_split(pDF$ID, "\\|", simplify = T)[, 2])
  pDF$UNIPROT <- id
  annoDf <-
    left_join(pDF, annoDf, by = c("UNIPROT" = "uniprotsptrembl"))
  return(annoDf)
}

SvsCTRL <-
  addAnnotation("Protein with different abundance S vs CTRL.csv")
write.csv(SvsCTRL,
          "Protein with different abundance with annotation S vs CTRL.csv",
          row.names = F)
COVIDvsCTRL <-
  addAnnotation("Protein with different abundance COVID vs CTRL.csv")
write.csv(
  COVIDvsCTRL,
  "Protein with different abundance with annotation COVID vs CTRL.csv",
  row.names = F
)
COVIDvsS <-
  addAnnotation("Protein with different abundance COVID vs S.csv")
write.csv(COVIDvsS,
          "Protein with different abundance with annotation COVID vs S.csv",
          row.names = F)
```


# get protein annotation from OrgDb

[Easy way to turn OrgDb object from AnnotationHub into package](https://support.bioconductor.org/p/92004/)

```{r}
#### make orgdb ####
# refreshHub(hubClass="AnnotationHub")
ah <- AnnotationHub()
# job::job({display(ah)})
# Chlorocebus sabaeus Green monkey
gm <- ah[["AH91675"]]
file.copy(gm, "./org.greenMonkey.eg.sqlite")

# seed <-
#   new(
#     "AnnDbPkgSeed",
#     Package = "org.Ng.eg.db",
#     Version = "0.0.1",
#     Author = "James W. MacDonald",
#     Maintainer = "James W. MacDonald <jmacdon@uw.edu>",
#     PkgTemplate = "NOSCHEMA.DB",
#     AnnObjPrefix = "org.Ng.eg",
#     organism = "Nannospalax galili",
#     species = "Nannospalax galili",
#     biocViews = "annotation",
#     manufacturerUrl = "none",
#     manufacturer = "none",
#     chipName = "none"
#   )
# makeAnnDbPkg(seed, "./org.greenMonkey.eg.sqlite")

con = dbConnect(SQLite(), "org.greenMonkey.eg.sqlite")
dbListTables(con)
gm = dbGetQuery(con,"select * from METADATA")
gm = dbGetQuery(con,"select * from DATA")
```


