---
title: "Learn LDpred"
author: "Yi Xiong"
date: "2021/3/24"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# ldpred2

[ldpred2 Tutorial](https://choishingwan.github.io/PRS-Tutorial/ldpred/)
[ldpred2 home page](https://github.com/privefl/bigsnpr)

```{r}

```

# ldpred

[ldpred home page](https://github.com/bvilhjal/ldpred)
[keyError: 'raw_snps'](https://github.com/bvilhjal/ldpred/issues/151)

## merge files

[If a line contains exactly three names, they are assumed to be the full filenames for a binary fileset (.bed, then .bim, then .fam).](https://www.cog-genomics.org/plink/1.9/data#merge_list)

```{r}
chr<-paste0("/work/biobank/ukb_data/1000G_3ph/all/bed_bim_fam/chr",1:22)
chrList<-data.frame(bim=paste0(chr,".bed"),
                    fam=paste0(chr,".bim"),
                    bed=paste0(chr,".fam"))
write.table(chrList,"ChrMergeList.txt",sep=" ",col.names = F,row.names = F,quote=F)
```

```{bash}
conda activate gwas
cp /work/biobank/ukb_data/scripts/merge.batch /scratch/yixiong
cd /scratch/yixiong/ldpred_yixiong
nohup plink \
  --merge-list ChrMergeList.txt \
  --make-bed \
  --out all \
  > ChrMerge.log.txt 2>&1 &   
```

## filter autosome of ukb db

X    X chromosome                    -> 23
Y    Y chromosome                    -> 24
XY   Pseudo-autosomal region of X    -> 25
MT   Mitochondrial                   -> 26

[plink data formats](https://zzz.bwh.harvard.edu/plink/data.shtml)

```{bash}
nohup plink \
  --bfile /work/biobank/ukb_data/diabetes/train \
  --autosome \
  --make-bed \
  --out train.autosome \
  > filterUKBautosome.log.txt 2>&1 &
```

```{r "filter autosome in p800k.csv"}
df<-read.csv("p800k.csv",sep="\t")
table(df$CHR)
df2<-df[df$CHR %in% 1:22,]
write.table(df2,"p800k.autosome.tsv",quote = F,sep="\t",row.names = F)
```

## QC

```{bash}
nohup plink \
  --threads 20 \
  --bfile all \
  --keep-allele-order \
  --me 1 1 \
  --set-me-missing \
  --make-bed \
  --out all1000g.nomendel \
  > all1000gQC.log.txt 2>&1 &

nohup plink \
  --bfile all1000g.nomendel \
  --keep-allele-order \
  --geno 0.05 \
  --mind 0.05 \
  --maf 0.01 \
  --hwe 1e-6 \
  --make-bed \
  --out all1000g.filter \
  > all1000gQC2.log.txt 2>&1 &
```

## coord

--gf LD Reference Genotype File.
--ssf Summary Statistic File.
--N Number of Individuals in Summary Statistic File.
--vgf Validation genotype file.

```{bash "IndexError: pop from empty list"}
nohup ldpred coord \
  --gf all \
  --ssf /work/biobank/ukb_data/diabetes/p800k.csv \
  --ssf-format CUSTOM \
  --N 500000 \
  --vgf train.autosome \
  --out coord.xy.result \
  > coord.xy.log.txt 2>&1 & 
```

swap reference and validation

```{bash "Unable to find phenotype values. "}

# error like "Unable to open object (object 'chrom_1' doesn't exist)"

nohup ldpred coord \
  --gf train.autosome \
  --ssf /work/biobank/ukb_data/diabetes/p800k.csv \
  --ssf-format CUSTOM \
  --N 500000 \
  --vgf all \
  --out coord.xy.result \
  > coord.xy.log.txt 2>&1 & 
```

[LDpred demo](https://hpc.nih.gov/apps/LDpred.html)

```{bash "use ukb autosome"}
ldpred coord \
  --gf train.autosome \
  --ssf p800k.autosome.tsv \
  --N 500000 \
  --ssf-format CUSTOM \
  --out coord.hdf5 \
  > coord.hdf5.log.txt 2>&1 & 
```


```{bash "use 1000 genome autosome"}

# Error like "Did not find chromosome 1 in SS dataset."

ldpred coord \
  --gf all \
  --ssf p800k.autosome.tsv \
  --N 500000 \
  --ssf-format CUSTOM \
  --out all1000g.coord.hdf5 \
  > all1000g.coord.hdf5.log.txt 2>&1 & 
```

```{bash "old test"}
conda activate gwas
pip install ldpred==1.0.10

cd /work/biobank/ukb_data/
nohup ldpred coord \
  --gf diabetes/train \
  --ssf p800k.csv \
  --ssf-format CUSTOM \
  --N 500000 \
  --vgf diabetes/train \
  --out step1.xy.result \
  > step1.xy.log.txt 2>&1 & 
  
nohup ldpred coord \
  --gf diabetes/train \
  --ssf p800k.csv \
  --ssf-format CUSTOM \
  --N 500000 \
  --vgf diabetes/train \
  --out /scratch/yixiong/ldpred_yixiong/step1.xy.result \
  > /scratch/yixiong/ldpred_yixiong/step1.xy.log.txt 2>&1 & 
```

## gibbs

```{bash}
# 784256
wc -l train.autosome.bim

# --ldr LDR LD radius. A value corresponding M/3000, where M is the number of 
# SNPs in the genome is recommended
# 784256/3000=261

# --N Number of individuals on which the summary statistics are assumed to be based on.

nohup ldpred gibbs \
  --cf coord.hdf5 \
  --ldf ukb.gibbs.ldf_file \
  --ldr 261 \
  --out ukb.gibbs \
  --N 500000 \
> ukb.gibbs.log.txt 2>&1 &
```

## score

```{bash "use 1000 genome autosome"}

# error assert num_individs > 0, 'No phenotypes were found!'

# --gf Validation genotype file.
# --rf RF SNP weights file (prefix used for output in preceding step)

nohup ldpred score \
  --gf all \
  --rf ukb.gibbs \
  --out all.score \
> all.score.log.txt 2>&1 &  
```

```{bash "use ukb"}

nohup ldpred score \
  --gf train.autosome \
  --rf ukb.gibbs \
  --out train.score \
> train.score.log.txt 2>&1 &  
```