---
title: "PRSice-2 workflow"
author: "xyz"
date: "2021/6/29"
output: 
  powerpoint_presentation:
    reference_doc: template.pptx
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(eval = TRUE)
```

# check summary statistics of diagram

  [UKB download page of genotype in plink format](http://biobank.ctsu.ox.ac.uk/crystal/label.cgi?id=263)

  [diagram data set](http://diagram-consortium.org/downloads.html)
  
  T2D GWAS meta-analysis
  Summary of T2D associations of HRC variants for only European UK Biobank subjects
  
  Mahajan.NatGenet2018b.UKBB.HRC.T2D.European.txt 
  [Read Me](http://diagram-consortium.org/downloads/Mahajan.et.al.2018b.European.GWAS.readme.pdf)
  
  
  
## Chromosome and position (build 37, base-pairs).
  
  o Effect (EA) and non-effect allele (NEA), aligned to the forward strand.
  
  o Effect allele frequency (EAF).
  
  o Log-odds ratio for the effect allele (Beta) and the corresponding standard error (SE). 
  
  o P-value for association (Pvalue).
  
  o Total reported effective sample size (Neff) / total sample size for UK Biobank (N).
  
  
  
## bim file format
  
  Chromosome code
  
  Position in morgans or centimorgans
  
  Base-pair coordinate
  
  Allele 1 (usually minor)
  
  Allele 2 (usually major)
  
# Search shared SNP between UKB and diagram

  13583103 SNPs in diagram
  
  no sex chromosome exists in diagram
  
  17642 SNP positions in diagram are duplicated
  
  704708 SNP positions of both sources are shared
  
  89.86% SNP in ukb can be found on diagram
  
  The shared SNP's position, p value, and odd ratio is in UKBdiagramShareSNP.tsv
  
  55140 SNP's pvalue is less than 0.05

```{r, eval=FALSE, include=FALSE}
library(data.table)
library(stringr)
library(dplyr)
# df<-fread("Mahajan.NatGenet2018b.UKBB.HRC.T2D.European.txt")
# saveRDS(df,"diagram.UKB.summaryStats.rds")
dfDiagram <- readRDS("diagram.UKB.summaryStats.rds")
dfTrain <- fread("train.autosome.bim.txt")
# 13583103 SNPs
dim(dfDiagram)
# no sex chromosome exists
unique(dfDiagram$Chr)
diagramPostion <- str_c(dfDiagram$Chr, ":", dfDiagram$Pos)
# SNP name is the same as diagramPostion
sum(diagramPostion != dfDiagram$SNP)
diagramPostion.duplicated <- duplicated(diagramPostion)
# 17642 SNP position is duplicated
sum(diagramPostion.duplicated)
diagramPostion.duplicated <-
  diagramPostion[diagramPostion.duplicated]
View(dfDiagram[dfDiagram$SNP == diagramPostion.duplicated[1]],)
View(dfDiagram[dfDiagram$SNP == diagramPostion.duplicated[2]],)
ukbPostion <- str_c(dfTrain$V1, ":", dfTrain$V4)
# 704708 positions are the same
samePos <- intersect(diagramPostion, ukbPostion)
length(samePos)
# 89.86% in ukb can be found on diagram
length(samePos) / nrow(dfTrain) * 100
dfDiagram$diagramPostion <- diagramPostion
dfTrain$ukbPostion <- ukbPostion
dfDiagramSame <-
  dfDiagram[match(samePos, dfDiagram$diagramPostion),]
dfTrainSame <- dfTrain[match(samePos, dfTrain$ukbPostion),]
dfTrainSammary <- fread("p800k.autosome.tsv")
colnames(dfTrainSammary)
# get the shared SNPs and P value
dfShare <- data.frame(
  CHR = dfTrainSame$V1,
  SNP = dfTrainSame$V2,
  BP = dfTrainSame$V4,
  A1 = dfTrainSame$V5,
  A2 = dfTrainSame$V6,
  OR = dfTrainSame$V3,
  P = dfDiagramSame$Pvalue
)
# 55140, in the paper is 50,224
# maybe some alleles are missed due to the same position
sum(dfShare$P <= 0.05)

#### get the unique match of SNP considering allele ####

dfDiagramSame <- dfDiagram[dfDiagram$diagramPostion %in% samePos, ]
dfDiagramSame <- arrange(dfDiagramSame, diagramPostion)
dfTrainSame <- dfTrain[dfTrain$ukbPostion %in% samePos, ]
dfTrainSame <- arrange(dfTrainSame, ukbPostion)
dfDiagramSame$diagramPostionAllele <-
  str_c(dfDiagramSame$diagramPostion,
        dfDiagramSame$EA,
        dfDiagramSame$NEA)
dfTrainSame$ukbPostionAllele1 <-
  str_c(dfTrainSame$ukbPostion, dfTrainSame$V5, dfTrainSame$V6)
dfTrainSame$ukbPostionAllele2 <-
  str_c(dfTrainSame$ukbPostion, dfTrainSame$V6, dfTrainSame$V5)
# 704336
sum((
  dfTrainSame$ukbPostionAllele1 %in% dfDiagramSame$diagramPostionAllele
) + (
  dfTrainSame$ukbPostionAllele2 %in% dfDiagramSame$diagramPostionAllele
)
)
P <- dfDiagramSame[, c(4, 5, 12, 7, 9)]
# convert odd ratio
P$Beta <- exp(P$Beta)
dfP <-
  left_join(dfTrainSame,
            P,
            by = c("ukbPostionAllele1" = "diagramPostionAllele"))
dfP <-
  left_join(dfP, P, by = c("ukbPostionAllele2" = "diagramPostionAllele"))
# no row is matched twice
unique(rowSums(is.na(dfP)))
P <- dfDiagramSame[, c(12, 7, 9)]
P$Beta <- exp(P$Beta)
dfP <-
  left_join(dfTrainSame,
            P,
            by = c("ukbPostionAllele1" = "diagramPostionAllele"))
dfP <-
  left_join(dfP, P, by = c("ukbPostionAllele2" = "diagramPostionAllele"))
dfP <- dfP[rowSums(is.na(dfP)) != 4, ]
dfP[is.na(dfP)] <- 0
dfP$Pvalue <- dfP$Pvalue.x + dfP$Pvalue.y
dfP$Beta <- dfP$Beta.x + dfP$Beta.y
dfP <- data.frame(
  CHR = dfP$V1,
  SNP = dfP$V2,
  BP = dfP$V4,
  A1 = dfP$V5,
  A2 = dfP$V6,
  OR = dfP$Beta,
  P = dfP$Pvalue
)
# 55111
sum(dfP$P <= 0.05)
write.table(
  dfP,
  "UKBdiagramShareSNP.tsv",
  sep = "\t",
  quote = F,
  row.names = F
)
```

# prsice demo

[tutorial](https://choishingwan.github.io/PRS-Tutorial/prsice/)

## required data

  file contains GWAS summary results
  
  *.bed	This file contains the genotype data that passed the QC steps
  
  *.bim	This file contains the list of SNPs that passed the QC steps
  
  *.fam	This file contains the samples that passed the QC steps
  
  file contains the phenotype data of the samples
  
  file contains the covariates of the samples (gender, for example)
  
  file contains the principal components (PCs) of the samples

```{r, eval=FALSE, include=FALSE}
library(data.table)
install.packages("optparse")
covariate <- fread("prsiceDemo/EUR.cov")
pcs <- fread("prsiceDemo/EUR.eigenvec", header=F)
colnames(pcs) <- c("FID","IID", paste0("PC",1:6))
cov <- merge(covariate, pcs)
fwrite(cov,"prsiceDemo/EUR.covariate", sep="\t")

demoDf<-fread("prsiceDemo/Height.QC")
shareDf<-fread("UKBdiagramShareSNP.tsv")
```

```{bash, eval=FALSE, include=FALSE}
cd prsiceDemo
Rscript bin/PRSice.R \
    --prsice bin/PRSice_linux \
    --base Height.QC.gz \
    --target EUR.QC \
    --binary-target F \
    --pheno EUR.height \
    --cov EUR.covariate \
    --base-maf MAF:0.01 \
    --base-info INFO:0.8 \
    --stat OR \
    --or \
    --out EUR
```

# QC

  --geno Exclude variants with missing call frequencies greater than a threshold (default 0.2)
  
  --mind Exclude samples with missing call frequencies greater than a threshold (default 0.2)
  
  --maf Exclude variants with minor allele frequency lower than a threshold (default 0.03)
  
  --hwe Exclude variants with Hardy-Weinberg equilibrium exact test p-values below 1e-6
  
## trainning and test data QC at the same time

### before filtering
  
  488377 samples (264797 females, 223467 males, 113 ambiguous; 488377 founders)
  
  784256 out of 805426 variants loaded
  
### after filtering
  
  488377 samples (264797 females, 223467 males, 113 ambiguous; 488377 founders)
  
  0 cases and 31700 controls
  
  remaining after main filters
  
  17930 variants removed due to missing genotype data
  
  198555 variants removed due to Hardy-Weinberg exact test
  
  230408 variants removed due to allele frequency threshold(s)
  
  337363 variants remaining after main filters

```{bash, eval=FALSE, include=FALSE}
# trainning and test data
cd /scratch/yixiong/prsice
nohup plink \
  --bfile /work/biobank/ukb_data/GenData/plink \
  --autosome \
  --make-bed \
  --out all.autosome \
  > filterUKBautosome.log.txt 2>&1 &
  
nohup plink2 \
  --threads 20 \
  --bfile /work/biobank/ukb_data/GenData/plink \
  --autosome \
  --geno 0.2 \
  --mind 0.2 \
  --maf 0.03 \
  --hwe 1e-6 \
  --make-bed \
  --out all.autosome.filter \
  > allQCfilter.log.txt 2>&1 &
```

## Only trainning data

```{bash, eval=FALSE, include=FALSE}
  
cd /scratch/yixiong/prsice
nohup plink \
  --threads 20 \
  --bfile train.autosome \
  --keep-allele-order \
  --me 1 1 \
  --set-me-missing \
  --make-bed \
  --out train.autosome.nomendel \
  > QCnomendel.log.txt 2>&1 &
  
nohup plink \
  --bfile /work/biobank/ukb_data/diabetes/train \
  --autosome \
  --make-bed \
  --out train.autosome \
  > filterUKBautosome.log.txt 2>&1 &

# --geno Exclude variants with missing call frequencies greater than a threshold (default 0.1)
# --mind Exclude samples with missing call frequencies greater than a threshold (default 0.1).
# --maf Exclude variants with minor allele frequency lower than a threshold (default 0.01).
# --hwe Exclude variants with Hardy-Weinberg equilibrium exact test p-values below a threshold.

nohup plink \
  --threads 20 \
  --bfile train.autosome.nomendel \
  --keep-allele-order \
  --geno 0.2 \
  --mind 0.2 \
  --maf 0.03 \
  --hwe 1e-6 \
  --make-bed \
  --out train.autosome.filter \
  > QCfilter.log.txt 2>&1 &
```

# PCA

[plink PCA tutorial](https://zhuanlan.zhihu.com/p/72490817)

 Get first 100 eigenvector
 
## all data

  useful results:

  all.PCA20.eigenval
  
  all.PCA20.eigenvec
  
  all.PCA20.log
  
  allPCA20.log.txt
  

```{bash, eval=FALSE, include=FALSE}
# memory is not enough
nohup plink2 \
  --threads 20 \
  --bfile all.autosome.filter \
  --pca approx 100 \
  --memory 250000 \
  --out all.autosome.filter.PCA \
  > allPCA.log.txt 2>&1 &

# get 10 PC

nohup plink2 \
  --threads 30 \
  --bfile all.autosome.filter \
  --pca approx 10 \
  --memory 250000 \
  --out all.PCA10 \
  > allPCA10.log.txt 2>&1 &

# get 20 PC

nohup plink2 \
  --threads 30 \
  --bfile all.autosome.filter \
  --pca approx 20 \
  --memory 250000 \
  --out all.PCA20 \
  > allPCA20.log.txt 2>&1 &
```

## training data

  train.autosome.filter.PCA20.eigenval
  
  train.autosome.filter.PCA20.eigenvec
  
  train.autosome.filter.PCA20.log

  trainPCA20.log.txt

```{bash, eval=FALSE, include=FALSE}
# Error: Failed to compute SVD of Krylov matrix (DGESVD info=20199)
nohup plink2 \
  --bfile train.autosome.filter \
  --pca approx 100 \
  --memory 204800 \
  --out train.autosome.filter.PCA \
  > PCA.log.txt 2>&1 &
  
nohup plink2 \
  --bfile train.autosome.filter \
  --pca approx 20 \
  --memory 250000 \
  --out train.autosome.filter.PCA20 \
  > trainPCA20.log.txt 2>&1 &
  
ls *PCA20* -l^C awk '{print $9}'
```

# Search shared SNP between UKB after QC and diagram

  13583103 SNPs in diagram
  
  337363 SNPs in UKB
  
  334433 positions are the same
  
  99.13% in ukb after QC can be found on diagram
   
  28773 SNPs p <= 0.05

```{r, eval=FALSE, include=FALSE}
library(data.table)
library(stringr)
library(dplyr)
dfDiagram <- readRDS("diagram.UKB.summaryStats.rds")
dfAll <- fread("all.autosome.filter.bim.txt")
# 13583103 SNPs in diagram
dim(dfDiagram)
# 337363 SNPs in UKB
nrow(dfAll)
diagramPostion <- str_c(dfDiagram$Chr, ":", dfDiagram$Pos)
ukbPostion <- str_c(dfAll$V1, ":", dfAll$V4)
# 334433 positions are the same
samePos <- intersect(diagramPostion, ukbPostion)
length(samePos)
# 99.13% in ukb after QC can be found on diagram
length(samePos) / nrow(dfAll) * 100
dfDiagram$diagramPostion <- diagramPostion
dfAll$ukbPostion <- ukbPostion

#### get the unique match of SNP considering allele ####

dfDiagramSame <- dfDiagram[dfDiagram$diagramPostion %in% samePos, ]
dfDiagramSame <- arrange(dfDiagramSame, diagramPostion)
dfAllSame <- dfAll[dfAll$ukbPostion %in% samePos, ]
dfAllSame <- arrange(dfAllSame, ukbPostion)
dfDiagramSame$diagramPostionAllele <-
  str_c(dfDiagramSame$diagramPostion,
        dfDiagramSame$EA,
        dfDiagramSame$NEA)
dfAllSame$ukbPostionAllele1 <-
  str_c(dfAllSame$ukbPostion, dfAllSame$V5, dfAllSame$V6)
dfAllSame$ukbPostionAllele2 <-
  str_c(dfAllSame$ukbPostion, dfAllSame$V6, dfAllSame$V5)
# 334347
sum((
  dfAllSame$ukbPostionAllele1 %in% dfDiagramSame$diagramPostionAllele
) + (
  dfAllSame$ukbPostionAllele2 %in% dfDiagramSame$diagramPostionAllele
)
)
P <- dfDiagramSame[, c(4, 5, 12, 7, 9)]
# convert odd ratio
P$Beta <- exp(P$Beta)
dfP <-
  left_join(dfAllSame,
            P,
            by = c("ukbPostionAllele1" = "diagramPostionAllele"))
dfP <-
  left_join(dfP, P, by = c("ukbPostionAllele2" = "diagramPostionAllele"))
# no row is matched twice
unique(rowSums(is.na(dfP)))
P <- dfDiagramSame[, c(12, 7, 9)]
P$Beta <- exp(P$Beta)
dfP <-
  left_join(dfAllSame,
            P,
            by = c("ukbPostionAllele1" = "diagramPostionAllele"))
dfP <-
  left_join(dfP, P, by = c("ukbPostionAllele2" = "diagramPostionAllele"))
dfP <- dfP[rowSums(is.na(dfP)) != 4, ]
dfP[is.na(dfP)] <- 0
dfP$Pvalue <- dfP$Pvalue.x + dfP$Pvalue.y
dfP$Beta <- dfP$Beta.x + dfP$Beta.y
dfP <- data.frame(
  CHR = dfP$V1,
  SNP = dfP$V2,
  BP = dfP$V4,
  A1 = dfP$V5,
  A2 = dfP$V6,
  OR = dfP$Beta,
  P = dfP$Pvalue
)
# 28773
sum(dfP$P <= 0.05)
write.table(
  dfP,
  "UKBdiagramShareSNPafterQC.tsv",
  sep = "\t",
  quote = F,
  row.names = F
)
```

# make input file on server

```{r eval=FALSE, include=FALSE}
library(data.table)
covariate <- fread("all.autosome.filter.fam")
all.diabetes <- covariate[, c(1, 2, 6)]
colnames(all.diabetes) <- c("FID", "IID", "Diabetes")
all.diabetes$Diabetes[all.diabetes$Diabetes == -9] <- 0
fwrite(all.diabetes, "all.diabetes", sep = "\t")
covariate <- covariate[, c(1, 2, 5)]
colnames(covariate) <- c("FID", "IID", "Sex")
pcs <- fread("all.PCA20.eigenvec", header = F, skip = 1)
pcs <- pcs[, 1:12]
colnames(pcs) <- c("FID", "IID", paste0("PC", 1:10))
cov <- merge(covariate, pcs)
fwrite(cov, "all.covariate", sep = "\t")
```

# run PRSice

```{bash, eval=FALSE, include=FALSE}
conda install -c conda-forge r-optparse
nohup Rscript PRSice.R \
    --prsice PRSice_linux \
    --base UKBdiagramShareSNPafterQC.tsv \
    --target all.autosome.filter \
    --binary-target T \
    --pheno all.diabetes \
    --cov all.covariate \
    --stat OR \
    --or \
    --thread 20 \
    --out allPR \
     > allPR.log.txt 2>&1 &
```



