---
title: "PRSice-2 workflow"
author: "xyz"
date: "2021/5/30"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# check summary statistics of diagram

[diagram data set] (http://diagram-consortium.org/downloads.html)

T2D GWAS meta-analysis
Summary of T2D associations of HRC variants for only European UK Biobank subjects

Mahajan.NatGenet2018b.UKBB.HRC.T2D.European.txt 
[Read Me](http://diagram-consortium.org/downloads/Mahajan.et.al.2018b.European.GWAS.readme.pdf)

Chromosome and position (build 37, base-pairs).
o Effect (EA) and non-effect allele (NEA), aligned to the forward strand.
o Effect allele frequency (EAF).
o Log-odds ratio for the effect allele (Beta) and the corresponding standard error (SE). 
o P-value for association (Pvalue).
o Total reported effective sample size (Neff) / total sample size for UK Biobank (N).

bim file format

Chromosome code
Position in morgans or centimorgans
Base-pair coordinate
Allele 1 (usually minor)
Allele 2 (usually major)

```{r}
library(data.table)
library(stringr)
library(dplyr)
# df<-fread("Mahajan.NatGenet2018b.UKBB.HRC.T2D.European.txt")
# saveRDS(df,"diagram.UKB.summaryStats.rds")
dfDiagram <- readRDS("diagram.UKB.summaryStats.rds")
dfTrain <- fread("train.autosome.bim.txt")
# 13583103 SNPs
dim(dfDiagram)
# no sex chromosome exists
unique(dfDiagram$Chr)
diagramPostion <- str_c(dfDiagram$Chr, ":", dfDiagram$Pos)
# SNP name is the same as diagramPostion
sum(diagramPostion != dfDiagram$SNP)
diagramPostion.duplicated <- duplicated(diagramPostion)
# 17642 SNP position is duplicated
sum(diagramPostion.duplicated)
diagramPostion.duplicated <-
  diagramPostion[diagramPostion.duplicated]
View(dfDiagram[dfDiagram$SNP == diagramPostion.duplicated[1]],)
View(dfDiagram[dfDiagram$SNP == diagramPostion.duplicated[2]],)
ukbPostion <- str_c(dfTrain$V1, ":", dfTrain$V4)
# 704708 positions are the same
samePos <- intersect(diagramPostion, ukbPostion)
length(samePos)
# 89.86% in ukb can be found on diagram
length(samePos) / nrow(dfTrain) * 100
dfDiagram$diagramPostion <- diagramPostion
dfTrain$ukbPostion <- ukbPostion
dfDiagramSame <-
  dfDiagram[match(samePos, dfDiagram$diagramPostion),]
dfTrainSame <- dfTrain[match(samePos, dfTrain$ukbPostion),]
dfTrainSammary <- fread("p800k.autosome.tsv")
colnames(dfTrainSammary)
# get the shared SNPs and P value
dfShare <- data.frame(
  CHR = dfTrainSame$V1,
  SNP = dfTrainSame$V2,
  BP = dfTrainSame$V4,
  A1 = dfTrainSame$V5,
  A2 = dfTrainSame$V6,
  OR = dfTrainSame$V3,
  P = dfDiagramSame$Pvalue
)
# 55140, in the paper is 50,224
# maybe some alleles are missed due to the same position
sum(dfShare$P <= 0.05)

#### get the unique match of SNP considering allele ####

dfDiagramSame <- dfDiagram[dfDiagram$diagramPostion %in% samePos, ]
dfDiagramSame <- arrange(dfDiagramSame, diagramPostion)
dfTrainSame <- dfTrain[dfTrain$ukbPostion %in% samePos, ]
dfTrainSame <- arrange(dfTrainSame, ukbPostion)
dfDiagramSame$diagramPostionAllele <-
  str_c(dfDiagramSame$diagramPostion,
        dfDiagramSame$EA,
        dfDiagramSame$NEA)
dfTrainSame$ukbPostionAllele1 <-
  str_c(dfTrainSame$ukbPostion, dfTrainSame$V5, dfTrainSame$V6)
dfTrainSame$ukbPostionAllele2 <-
  str_c(dfTrainSame$ukbPostion, dfTrainSame$V6, dfTrainSame$V5)
# 704336
sum((
  dfTrainSame$ukbPostionAllele1 %in% dfDiagramSame$diagramPostionAllele
) + (
  dfTrainSame$ukbPostionAllele2 %in% dfDiagramSame$diagramPostionAllele
)
)
P <- dfDiagramSame[, c(4, 5, 12, 7, 9)]
# convert odd ratio
P$Beta <- exp(P$Beta)
dfP <-
  left_join(dfTrainSame,
            P,
            by = c("ukbPostionAllele1" = "diagramPostionAllele"))
dfP <-
  left_join(dfP, P, by = c("ukbPostionAllele2" = "diagramPostionAllele"))
# no row is matched twice
unique(rowSums(is.na(dfP)))
P <- dfDiagramSame[, c(12, 7, 9)]
P$Beta <- exp(P$Beta)
dfP <-
  left_join(dfTrainSame,
            P,
            by = c("ukbPostionAllele1" = "diagramPostionAllele"))
dfP <-
  left_join(dfP, P, by = c("ukbPostionAllele2" = "diagramPostionAllele"))
dfP <- dfP[rowSums(is.na(dfP)) != 4, ]
dfP[is.na(dfP)] <- 0
dfP$Pvalue <- dfP$Pvalue.x + dfP$Pvalue.y
dfP$Beta <- dfP$Beta.x + dfP$Beta.y
dfP <- data.frame(
  CHR = dfP$V1,
  SNP = dfP$V2,
  BP = dfP$V4,
  A1 = dfP$V5,
  A2 = dfP$V6,
  OR = dfP$Beta,
  P = dfP$Pvalue
)
# 55111
sum(dfP$P <= 0.05)
write.table(
  dfP,
  "UKBdiagramShareSNP.tsv",
  sep = "\t",
  quote = F,
  row.names = F
)
```

# prsice demo

[tutorial](https://choishingwan.github.io/PRS-Tutorial/prsice/)

```{r}
library(data.table)
covariate <- fread("prsiceDemo/EUR.cov")
pcs <- fread("prsiceDemo/EUR.eigenvec", header=F)
colnames(pcs) <- c("FID","IID", paste0("PC",1:6))
cov <- merge(covariate, pcs)
fwrite(cov,"prsiceDemo/EUR.covariate", sep="\t")

demoDf<-fread("prsiceDemo/Height.QC")
shareDf<-fread("UKBdiagramShareSNP.tsv")
```

```{bash}
cd prsiceDemo
Rscript bin/PRSice.R \
    --prsice bin/PRSice_linux \
    --base Height.QC.gz \
    --target EUR.QC \
    --binary-target F \
    --pheno EUR.height \
    --cov EUR.covariate \
    --base-maf MAF:0.01 \
    --base-info INFO:0.8 \
    --stat OR \
    --or \
    --out EUR
```

# QC

```{bash}
nohup plink \
  --bfile /work/biobank/ukb_data/GenData/plink \
  --autosome \
  --make-bed \
  --out all.autosome \
  > filterUKBautosome.log.txt 2>&1 &
  
cd /scratch/yixiong/prsice
nohup plink \
  --threads 20 \
  --bfile train.autosome \
  --keep-allele-order \
  --me 1 1 \
  --set-me-missing \
  --make-bed \
  --out train.autosome.nomendel \
  > QCnomendel.log.txt 2>&1 &
  
nohup plink \
  --bfile /work/biobank/ukb_data/diabetes/train \
  --autosome \
  --make-bed \
  --out train.autosome \
  > filterUKBautosome.log.txt 2>&1 &

# --geno Exclude variants with missing call frequencies greater than a threshold (default 0.1)
# --mind Exclude samples with missing call frequencies greater than a threshold (default 0.1).
# --maf Exclude variants with minor allele frequency lower than a threshold (default 0.01).
# --hwe Exclude variants with Hardy-Weinberg equilibrium exact test p-values below a threshold.

nohup plink \
  --threads 20 \
  --bfile train.autosome.nomendel \
  --keep-allele-order \
  --geno 0.2 \
  --mind 0.2 \
  --maf 0.03 \
  --hwe 1e-6 \
  --make-bed \
  --out train.autosome.filter \
  > QCfilter.log.txt 2>&1 &
```

# PCA

[plink PCA tutorial](https://zhuanlan.zhihu.com/p/72490817)

```{bash}
nohup plink2 \
  --bfile train.autosome.filter \
  --pca approx 10 \
  --memory 204800 \
  --out train.autosome.filter.PCA \
  > PCA.log.txt 2>&1 &
```


