---
title: "PRSice-2 workflow"
author: "xyz"
date: "2021/5/30"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# check summary statistics of diagram

[diagram data set] (http://diagram-consortium.org/downloads.html)

T2D GWAS meta-analysis
Summary of T2D associations of HRC variants for only European UK Biobank subjects

Mahajan.NatGenet2018b.UKBB.HRC.T2D.European.txt Read Me

Chromosome and position (build 37, base-pairs).
o Effect (EA) and non-effect allele (NEA), aligned to the forward strand.
o Effect allele frequency (EAF).
o Log-odds ratio for the effect allele (Beta) and the corresponding standard error (SE). 
o P-value for association (Pvalue).
o Total reported effective sample size (Neff) / total sample size for UK Biobank (N).

bim file format

Chromosome code
Position in morgans or centimorgans
Base-pair coordinate
Allele 1 (usually minor)
Allele 2 (usually major)

```{r}
library(data.table)
library(stringr)
library(dplyr)
# df<-fread("Mahajan.NatGenet2018b.UKBB.HRC.T2D.European.txt")
# saveRDS(df,"diagram.UKB.summaryStats.rds")
df <- readRDS("diagram.UKB.summaryStats.rds")
df2 <- fread("train.autosome.bim.txt")
# 13583103 SNPs
dim(df)
# no sex chromosome exists
unique(df$Chr)
diagramPostion <- str_c(df$Chr, ":", df$Pos)
# SNP name is the same as diagramPostion
sum(diagramPostion!=df$SNP)
diagramPostion.duplicated <- duplicated(diagramPostion)
# 17642 SNP position is duplicated
sum(diagramPostion.duplicated)
diagramPostion.duplicated <- diagramPostion[diagramPostion.duplicated]
View(df[df$SNP == diagramPostion.duplicated[1]], )
View(df[df$SNP == diagramPostion.duplicated[2]], )
ukbPostion <- str_c(df2$V1, ":", df2$V4)
# 704708 positions are the same
samePos <- intersect(diagramPostion, ukbPostion)
length(samePos)
# 89.86% in ukb can be found on diagram
length(samePos) / nrow(df2) * 100
df$diagramPostion <- diagramPostion
df2$ukbPostion <- ukbPostion
dfSame <- df[match(samePos, df$diagramPostion), ]
df2Same <- df2[match(samePos, df2$ukbPostion), ]
df3<-fread("p800k.autosome.tsv")
colnames(df3)
# get the shared SNPs and P value
df4<-data.frame(CHR=df2Same$V1,
                SNP=df2Same$V2,
                BP=df2Same$V4,
                A1=df2Same$V5,
                A2=df2Same$V6,
                OR=df2Same$V3,
                P=dfSame$Pvalue)
# 55140, in the paper is 50,224
# maybe some alleles are missed due to the same position
sum(df4$P<=0.05)

#### get the unique match of SNP considering allele ####
dfSame2 <- df[df$diagramPostion %in% samePos,]
dfSame2 <- arrange(dfSame2, diagramPostion)
df2Same2 <- df2[df2$ukbPostion %in% samePos,]
df2Same2 <- arrange(df2Same2, ukbPostion)
dfSame2$diagramPostionAllele <-
  str_c(dfSame2$diagramPostion, dfSame2$EA, dfSame2$NEA)
df2Same2$ukbPostionAllele1 <-
  str_c(df2Same2$ukbPostion, df2Same2$V5, df2Same2$V6)
df2Same2$ukbPostionAllele2 <-
  str_c(df2Same2$ukbPostion, df2Same2$V6, df2Same2$V5)
# 704336
sum((df2Same2$ukbPostionAllele1 %in% dfSame2$diagramPostionAllele) + (df2Same2$ukbPostionAllele2 %in% dfSame2$diagramPostionAllele)
)
p<-dfSame2[,c(4,5,12,9)]
dfP<-left_join(df2Same2,p,by=c("ukbPostionAllele1"="diagramPostionAllele"))
dfP<-left_join(dfP,p,by=c("ukbPostionAllele2"="diagramPostionAllele"))
# no row is matched twice
unique(rowSums(is.na(dfP)))
p<-dfSame2[,c(12,9)]
dfP<-left_join(df2Same2,p,by=c("ukbPostionAllele1"="diagramPostionAllele"))
dfP<-left_join(dfP,p,by=c("ukbPostionAllele2"="diagramPostionAllele"))
dfP<-dfP[rowSums(is.na(dfP))!=2,]
dfP[is.na(dfP)]<-0
dfP$Pvalue<-dfP$Pvalue.x+dfP$Pvalue.y
dfP2<-data.frame(CHR=dfP$V1,
                SNP=dfP$V2,
                BP=dfP$V4,
                A1=dfP$V5,
                A2=dfP$V6,
                OR=dfP$V3,
                P=dfP$Pvalue)
# 55111
sum(dfP2$P<=0.05)
write.table(dfP2,"UKBdiagramShareSNP.tsv",sep = "\t",quote = F,row.names = F)
```

# prsice demo

[tutorial](https://choishingwan.github.io/PRS-Tutorial/prsice/)

```{r}
library(data.table)
covariate <- fread("prsiceDemo/EUR.cov")
pcs <- fread("prsiceDemo/EUR.eigenvec", header=F)
colnames(pcs) <- c("FID","IID", paste0("PC",1:6))
cov <- merge(covariate, pcs)
fwrite(cov,"prsiceDemo/EUR.covariate", sep="\t")
```

```{bash}
cd prsiceDemo
Rscript bin/PRSice.R \
    --prsice bin/PRSice_linux \
    --base Height.QC.gz \
    --target EUR.QC \
    --binary-target F \
    --pheno EUR.height \
    --cov EUR.covariate \
    --base-maf MAF:0.01 \
    --base-info INFO:0.8 \
    --stat OR \
    --or \
    --out EUR
```

